cmd/libs/go2idl/parser/parse_test.go

```diff
@@ -0,0 +1,350 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser_test
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+	"text/template"
+
+	"k8s.io/kubernetes/cmd/libs/go2idl/namer"
+	"k8s.io/kubernetes/cmd/libs/go2idl/parser"
+	"k8s.io/kubernetes/cmd/libs/go2idl/types"
+)
+
+func construct(t *testing.T, files map[string]string, testNamer namer.Namer) (*parser.Builder, types.Universe, []*types.Type) {
+	b := parser.New()
+	for name, src := range files {
+		if err := b.AddFile(name, []byte(src)); err != nil {
+			t.Fatal(err)
+		}
+	}
+	u, err := b.FindTypes()
+	if err != nil {
+		t.Fatal(err)
+	}
+	orderer := namer.Orderer{testNamer}
+	o := orderer.Order(u)
+	return b, u, o
+}
+
+func TestBuilder(t *testing.T) {
+	var testFiles = map[string]string{
+		"base/foo/proto/foo.go": `
+package foo
+
+import (
+	"base/common/proto"
+)
+
+type Blah struct {
+	common.Object
+	Count int64
+	Frobbers map[string]*Frobber
+	Baz []Object
+	Nickname *string
+	NumberIsAFavorite map[int]bool
+}
+
+type Frobber struct {
+	Name string
+	Amount int64
+}
+
+type Object struct {
+	common.Object
+}
+
+`,
+		"base/common/proto/common.go": `
+package common
+
+type Object struct {
+	ID int64
+}
+`,
+	}
+
+	var tmplText = `
+package o
+{{define "Struct"}}type {{Name .}} interface { {{range $m := .Members}}{{$n := Name $m.Type}}
+	{{if $m.Embedded}}{{$n}}{{else}}{{$m.Name}}() {{$n}}{{if $m.Type.Elem}}{{else}}
+	Set{{$m.Name}}({{$n}}){{end}}{{end}}{{end}}
+}
+
+{{end}}{{define "Map"}}{{end}}{{define "Slice"}}{{end}}
+{{range $t := .}}{{if eq $t.Kind "Struct"}}{{template "Struct" $t}}{{end}}{{end}}`
+
+	var expect = `
+package o
+
+type CommonObject interface { 
+	ID() Int64
+	SetID(Int64)
+}
+
+type FooBlah interface { 
+	CommonObject
+	Count() Int64
+	SetCount(Int64)
+	Frobbers() MapStringToPointerFooFrobber
+	Baz() SliceFooObject
+	Nickname() PointerString
+	NumberIsAFavorite() MapIntToBool
+}
+
+type FooFrobber interface { 
+	Name() String
+	SetName(String)
+	Amount() Int64
+	SetAmount(Int64)
+}
+
+type FooObject interface { 
+	CommonObject
+}
+
+`
+	testNamer := namer.NewPublicNamer(1, "proto")
+	_, u, o := construct(t, testFiles, testNamer)
+	t.Logf("\n%v\n\n", o)
+	tmpl := template.Must(
+		template.New("").
+			Funcs(
+			map[string]interface{}{
+				"Name": testNamer.Name,
+			}).
+			Parse(tmplText),
+	)
+	buf := &bytes.Buffer{}
+	tmpl.Execute(buf, o)
+	if e, a := expect, buf.String(); e != a {
+		t.Errorf("Wanted, got:\n%v\n-----\n%v\n", e, a)
+	}
+
+	testNamer = namer.NewPrivateNamer(0)
+	testNamer.IgnoreWords = map[string]bool{"proto": true}
+	orderer := namer.Orderer{testNamer}
+	o = orderer.Order(u)
+	t.Logf("\n%v\n\n", o)
+
+	// TODO: enable this when packge import graph works
```

Comment:
Argh, after some work-- yes. (didn't work if you added single files, as the tests all do...)